platform/search/src/services/GenericSearchProvider.js

```diff
@@ -110,83 +125,50 @@ define(
-                        }, 0);
-                    }
-                    
-                    // Watch for changes to this item, in case it gets new children 
-                    if (node && node.hasCapability && node.hasCapability('mutation')) {
-                        node.getCapability('mutation').listen(function (listener) {
-                            if (listener && listener.composition) {
-                                // If the node was mutated to have children, get the child domain objects
-                                objectService.getObjects(listener.composition).then(function (objectsById) {
-                                    var objects = [],
-                                        id;
 
-                                    // Get each of the domain objects in objectsById
-                                    for (id in objectsById) {
-                                        objects.push(objectsById[id]);
-                                    }
-                                    
-                                    indexItems(objects);
-                                });
-                            }
-                        });
-                    }
-                });
-            }
-            
-            // Converts the filetree into a list
-            function getItems() {
-                // Aquire root objects
-                objectService.getObjects(ROOTS).then(function (objectsById) {
-                    var objects = [],
-                        id;
-                    
-                    // Get each of the domain objects in objectsById
-                    for (id in objectsById) {
-                        objects.push(objectsById[id]);
+            function requestAndIndex(id) {
+                delete toRequest[id];
+                pendingRequests += 1;
+                objectService.getObjects([id]).then(function (objects) {
+                    if (objects[id]) {
+                        indexItem(objects[id]);
                     }
-                    
-                    // Index all of the roots' descendents
-                    indexItems(objects);
+                }, function () {
+                    $log.warn("Failed to index domain object " + id);
+                }).then(function () {
+                    pendingRequests -= 1;
+                    scheduleFlush();
                 });
             }
 
+            scheduleFlush = throttle(function flush() {
+                var batchSize =
+                        Math.max(MAX_CONCURRENT_REQUESTS - pendingRequests, 0);
+
+                if (Object.keys(toRequest).length + pendingRequests < 1) {
+                    $log.info([
+                        'GenericSearch finished indexing after ',
+                        ((Date.now() - indexingStarted) / 1000).toFixed(2),
+                        ' seconds.'
+                    ].join(''));
+                } else {
+                    Object.keys(toRequest)
```

Comment:
Seems like toRequest should not be an object: you're constantly coercing it to an array (spendy) and never using it to check for membership. it would perform better as an array.